sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py

```diff
@@ -280,6 +283,18 @@ def test_max_depth(max_depth):
     assert depth == max_depth
 
 
+def test_max_depth_max_leaf_nodes():
```

Comment:
You don't think that we should move this test in `test_gradient_boosting.py` instead of the `test_grower.py`?

Reply:
No strong opinion. It's really a test about the grower code more than the estimator but yeah it's not obvious to come up with something that would only involve the grower

Reply:
Okay this may seem a little crazy, but:

```py
def test_max_depth_max_leaf_nodes():
    n_samples = 1000
    n_bins = 64
    max_depth = 4
    max_leaf_nodes = 9  # 2**(max_depth - 1) + 1

    X, y = make_classification(n_samples=n_samples, random_state=42)
    X_binned = _BinMapper(n_bins=n_bins, random_state=42).fit_transform(X)
    y = y.astype(Y_DTYPE)

    loss = LeastSquares()
    baseline = loss.get_baseline_prediction(y, 1)
    raw_predictions = np.full(shape=(1, n_samples), fill_value=baseline)

    gradients, hessians = loss.init_gradients_and_hessians(
        n_samples=n_samples, prediction_dim=1)

    for i in range(5):
        loss.update_gradients_and_hessians(gradients, hessians, y,
                                           raw_predictions)
        grower = TreeGrower(X_binned, gradients[0, :], hessians[0, :],
                            n_bins=n_bins, max_leaf_nodes=max_leaf_nodes,
                            max_depth=max_depth)
        grower.grow()
        _update_raw_predictions(raw_predictions[0, :], grower)
        assert len(grower.finalized_leaves) <= max_leaf_nodes

```